### PR TITLE
Fix PP not being calculated correctly for nightcore scores

### PR DESCRIPTION
**Pull request by bdach**
- Closes https://github.com/ppy/osu/issues/25175

I guess this would be one way to roll forward with https://github.com/ppy/osu/pull/24640, but there is a chance something somewhere else could have relied on the inheritance and might now break, so not sure whether this should be the solution here...

Check is slightly rewritten since it wasn't making much sense to me previously (why add nightcore to allowed mods if double time is selected?)

### CODE REVIEW

**Comment by peppy** [comment]
Can I get an ELI5 of how this fixes things?

**Comment by bdach** [comment]
`CreateDifficultyModCombinations()` will generally only output combinations with double time and ignore nightcore, on the basis that the two are interchangeable. But these two methods don't generally want to drop nightcore, so the logic goes that if a score has nightcore on it, add it to the set of allowed mods, as it wouldn't generally be there due to the aforementioned method ignoring its existence.

This broke because we removed the `ModNightcore : ModDoubleTime` inheritance.

**Comment by peppy** [comment]
To confirm, `IsSubclassOf` is basically a synonym for what was already there, right?

**Comment by smoogipoo** [comment]
One thing that could be considered here is to take [this](https://github.com/ppy/osu-queue-score-statistics/blob/master/osu.Server.Queues.ScoreStatisticsProcessor/Helpers/LegacyModsHelper.cs) implementation and do a conversion of `Mod[] -> Ruleset.ConvertToLegacyMods() -> MaskRelevantMods() -> ConvertFromLegacyMods()`

The reasoning is that the `MaskRelevantMods()` method is used for looking up database attributes, so it must match one of the values of `DifficultyCalculator.DifficultyAdjustmentMods()`.

I'd even suggest moving that to the base game project.

**Comment by bdach** [comment]
Not really. That part is checking if double time is in `allowedMods` at all. It doesn't really correspond to anything in the old code, as it just seemed to assume double time will always be in `allowedMods` and just blindly added nightcore whenever encountering it (...or double time, which is similarly weird).

**Comment by peppy** [comment]
> Not really. That part is checking if double time is in allowedMods at all

Right, I missed the variable change at the start...

**Review by peppy — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]
Silly GitHub... I had a pending review comment, but it's not critical.

File: PerformanceCalculator/LegacyHelper.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ public static Mod[] ConvertToLegacyDifficultyAdjustmentMods(BeatmapInfo beatmapI
                                       .ToHashSet();
 
             // Special case to allow either DT or NC.
-            if (mods.Any(m => m is ModDoubleTime))
+            if (allowedMods.Any(type => type.IsSubclassOf(typeof(ModDoubleTime))) && mods.Any(m => m is ModNightcore))
                 allowedMods.Add(allMods.Single(m => m is ModNightcore).GetType());
```
[the remark this answers is not in the document]
Comment (bdach) [reply PerformanceCalculator/LegacyHelper.cs:82]: ...by adding nightcore to `DifficultyAdjustmentMods` game-side, I presume? I'm not sure how you'd preserve behaviour otherwise.

[the remark this answers is not in the document]
Comment (smoogipoo) [reply PerformanceCalculator/LegacyHelper.cs:82]: I was thinking it could be added unconditionally to `allowedMods` here?

[the remark this answers is not in the document]
Comment (bdach) [reply PerformanceCalculator/LegacyHelper.cs:82]: Hmmmmmmmm I think I prefer what's already here. You're probably right that unconditionally adding wouldn't probably ever be a problem but it is what it is.
